hash_practice/exercises.py

```diff
@@ -1,29 +1,164 @@
 
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
-        Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Each subarray will have strings which are anagrams of each other.
+        Time Complexity: O(n^2) -- or is it O(n) bc of the "continue"s?
+        Space Complexity: O(n)
     """
-    pass
+    result = []
+
+    for i in range(len(strings)):
+        if strings[i] == None:
+            continue
+
+        anagrams = [strings[i]]
+        strings[i] = None
+
+        for j in range(len(strings)):
+            if strings[j] == None:
+                continue
+
+            if __is_anagram(anagrams[0], strings[j]):
+                anagrams.append(strings[j])
+                strings[j] = None
+
+        result.append(anagrams)
+
+    return result
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n x m), where n is length of original list and m is number of repeats for each repeated value
+        Space Complexity: O(n)
     """
-    pass
+    top_k = []
+
+    freqs = __create_indexed_freq_map(nums)
+    reverse_freqs = __create_reverse_indexed_freq_map(freqs)
+
+    print(reverse_freqs)
+
+    sorted_freqs = sorted(reverse_freqs.keys(), reverse=True)
+    
+    for freq in sorted_freqs:
+        if k - len(top_k) >= len(reverse_freqs[freq]):
+            for index_num in reverse_freqs[freq]:
+                top_k.append(index_num[1])
+        else:
+            sorted_index_nums = sorted(reverse_freqs[freq])
+            for index_num in sorted_index_nums:
+                top_k.append(index_num[1])
+                if len(top_k) == k:
+                    return top_k
+        
+        if len(top_k) == k:
+            return top_k
+
+    return top_k
 
 
-def valid_sudoku(table):
+def valid_sudoku(board):
     """ This method will return the true if the table is still
         a valid sudoku table.
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2)
+        Space Complexity: O(n) ? its a little complicated and im tired

```

Comment:
Just a note that the time complexity is really O(1) since the size of the input never changes.  The grid is always 9x9.